### DISCUSSION TITLE
Aggregating data for historical purposes

### DISCUSSION THREAD

**Discussion by ViktorJp** [vloschiavo/powerwall2#79] [Q&A] [answered]
Again, thank you very much for all this awesome documentation out on your github page, and for answering my last question about grid outages.  I was wondering if I could please get your wisdom on this issue:

- Like with the Tesla app, you are able to see a plethora of historical power information over the past 24hrs, days, weeks, months, etc.  I'm guessing that a constant stream of data is heading over to Tesla, and they are aggregating all that info there to display on their apps.  I'm not seeing anything in this API documentation that would give me the same kinds of historical power info (other than the total lifetime kwH going to/from Grid I believe). 
- My question is, what is the best practice in gathering this data, storing and calculating it without the means of being able to use a database?  I was going to take periodic snapshots (like every 10s), add it up and average it out over a total span of time for each piece - grid, home, pw and solar, and saving/reading this from a file.  I know it probably won't be as accurate as what we can get from the Tesla site stats.  Is there a better way I'm not seeing?

Would love your input!  :)  Thank you!

**Comment by GBDownUnder** [comment]
You can download the historic data from Tesla via the Tesla app.  I would think just doing this occasionally would be your best option.

From: ViktorJp ***@***.***>
Sent: Tuesday, 16 August 2022 7:29 AM
To: vloschiavo/powerwall2 ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [vloschiavo/powerwall2] Aggregating data for historical purposes (Discussion #79)


Again, thank you very much for all this awesome documentation out on your github page, and for answering my last question about grid outages. I was wondering if I could please get your wisdom on this issue:

  *   Like with the Tesla app, you are able to see a plethora of historical power information over the past 24hrs, days, weeks, months, etc. I'm guessing that a constant stream of data is heading over to Tesla, and they are aggregating all that info there to display on their apps. I'm not seeing anything in this API documentation that would give me the same kinds of historical power info (other than the total lifetime kwH going to/from Grid I believe).
  *   My question is, what is the best practice in gathering this data, storing and calculating it without the means of being able to use a database? I was going to take periodic snapshots (like every 10s), add it up and average it out over a total span of time for each piece - grid, home, pw and solar, and saving/reading this from a file. I know it probably won't be as accurate as what we can get from the Tesla site stats. Is there a better way I'm not seeing?

Would love your input! :) Thank you!

**Reply by ViktorJp** [reply]
OK... but what if you don't want to go the Tesla app route, and all you have is the gateway API to play with... ie. accumulating stats through that.  What's the simplest way to go?  My object is to build a simple method that doesn't rely on a database that calculates stats over time for display purposes using the gateway API as the only source.

**Reply by chrisheninger** [reply]
From what I've seen, a lot of folks use Home Assistant with the Powerwall plugin to record data in the way that you are describing. (And since HA is all open-source, you should be able to see how they implemented it and even access the data directly)

Edit: here's a link: https://www.home-assistant.io/integrations/powerwall/

**Comment by GBDownUnder** [comment, marked as answer]
If you don’t want to go the Telsa app route then sure, you will need to do all that coding and data analysis work yourself.  It will be less accurate as you can only sample at a rate which will always be lower than the Powerwall can internally, but that’s your only option.
The gateway API does provide some aggregates but there are no details I know of that detail what time frame they are calculated over.

**Reply by ViktorJp** [reply]
Thank you for validating... I expected to get this answer.  ;)
